# Purpose: Analysis of cfDNA fragment coverage of patient data.
# Author: Yasamin Nouri Jelyani
# Concept from: Jonathan Broadbent
# Date: 2022-12-20
# Version: 0.1.0
# Bugs and Issues: Takes too long to run


#------------------------Exported Function------------------------



#' Nucleosome coverage plots for chromosome positions. Used to find Transcription
#' Factor Binding Sites (TFBS)
#'
#'TFBS can be used to subtype cancer DNA. Transcription Factors (TFs)bind to
#'specific regions of specific kinds of DNA, to allow for their transcription.
#'The binding of the TFs allow for generation of RNA from that DNA region.
#'It is important to note that regions that are binded by TFs are not bound
#'by nucleosomes (to allow for binding of the TF protein). However, other
#'regions of the DNA are bound by these nucleosomes, protecting the DNA from
#'degradation. cfDNA samples that do not have TF bound, do not have
#'nucleosomes bound at the region where TF usually binds. Hence, they are
#'not protected from nuclease degradation. This lead to the cfDNA to be cut at
#'the TF binding sites since those sites are unprotected. Hence, cfDNA fragment
#'ends if many of the cfDNA are cut at that certain region can be used as an
#'indicator for TF binding sites. Hence, coverage plots are used to detect the
#'amount of coverage by the cfDNA molecules on the DNA. If the coverage is low,
#'that could indicate the existance of a TFBS in that region.
#'The coverage plots are created using the nucleosomeCoverage data structure.
#'
#'
#' Note 1: The data to generate examples are from the illumina website (4).
#'  This example data is not representative of real patient data due to
#'  limitations to access of real patient data.
#'  However, users can use their own real patient datasets for
#'  analysis.
#'  More examples can be found at https://github.com/Yasamin-Nourijelyani/CfDNAfragmentomics/tree/master/inst/extdata
#'
#' Note 2:
#' BED files can be generate from BAM files using bedtools bamtobed function
#' in the command line (3).
#'
#' Since BED files are commonly
#' generated by computational biologists and are easily convertible, from
#' BAM files, they can be used as input. Otherwise, tab seperated
#' .txt files that generate data in the same format of bed files
#' with the same columns as described in the parameters are also
#' excepted.
#'
#'
#' @param sample_bed A dataframe for the a sample data
#' (unknown if cancerous or not individual)
#'  with at least 3 columns: first column is a string representing the chromosome
#'  number for the cfDNA for instance: "chr1",
#'  the second column are integer values indicating the start location of
#'  the read and third column are integer values indicating end location of
#'  the read.
#' These are the patient who we want to check if they have cancer or not.
#' The fragment lengths
#' of patients will be compared to these healthy data similar to what is done in
#' the paper by Katsman et. al (1).
#'
#'
#' #TODO:
#' add references to content above.
#'
#' @return Return an dataframe of nucleosome coverage:
#' value refer to the number of cfDNA fragments that cover the region of the
#' nucleosome.
#'
#' @references
#' 1- Katsman, E., Orlanski, S., Martignano, F., Fox-Fisher, I., Shemer,
#'  R., Dor, Y., Zick, A.,
#' Eden, A., Petrini, I., Conticello, S. G., & Berman, B. P. (2022).
#' Detecting cell-of-origin and cancer-specific methylation features of
#' cell-free DNA from Nanopore sequencing.
#' Genome biology, 23(1), 158.
#' https://doi.org.myaccess.library.utoronto.ca/10.1186/s13059-022-02710-1
#'
#' 2- “Mann Whitney U Test in R Programming.” GeeksforGeeks, 28 Dec. 2021,
#' https://www.geeksforgeeks.org/mann-whitney-u-test-in-r-programming/.
#'
#' 3- Quinlan, Aaron R., Kindlon, Neil. "bamtobed" readthedocs, May 26, 2022,
#'  https://bedtools.readthedocs.io/en/latest/content/tools/bamtobed.html
#'
#' 4- Illumina, https://support.illumina.com/downloads/nextera-flex-for-enrichment-BED-files.html
#'
#' 5- “The T-Test.” JMP,
#' https://www.jmp.com/en_ca/statistics-knowledge-portal/t-test.html#:~
#' :text=t%2DTest%20assumptions&amp;text=The%20data%20are%20continuous.,
#' The%20distribution%20is%20approximately%20normal.
#'
#'@examples
#' \dontrun{
#'
#'
#' sample_bed_file <- system.file("extdata", "d1.bed", package = "CfDNAfragmentomics")
#' sample_bed <- read.table(sample_bed_file, header = FALSE, sep="\t",stringsAsFactors=FALSE, quote="")
#'
#' ## basic usage
#'
#' # Example 1:
#'
#' cov <- nucleosomeCoverage(sample_bed = sample_bed)
#' cov
#'
#' }
#'
#'
#' @export


nucleosomeCoverage <- function(sample_bed){

  #check if the input values are correct.
  if (! is.data.frame(sample_bed)) {
    stop("Please put valid inputs for nucleosomeCoverage function")
  }

  # the maximum length of cfDNA in this file
  chromosome_length = as.numeric(max(sample_bed[, 3]) - min(sample_bed[, 2]))
  # coverage array
  coverage = c(1:chromosome_length) * 0

  #row in bed file (1 cfDNA)
  for (i in c(1:nrow(sample_bed))){
      #traverse the length of the cfDNA fragment
      for (j in c(sample_bed[i, 2]:sample_bed[i, 3])){
        coverage[j] =  coverage[j] + 1

      }
    }




  # check for instance coverage[2985822]
  return(as.data.frame(coverage))


}




# [END]

